config.txt
Configuration Settings

INIT:  Initialize game state with predefined settings
  - asset_path ("assets"):  directory to load assets from
  - color_id (0):  ID of the color scheme to use
  - white_id (0):  ID of white starting army
  - black_id (0):  ID of black starting army
    - >= 0:  use predefined army
    - < 0:  use chaotic army
  - edit_id (None):  ID of army to place pieces from in edit mode
    - int:  see above
    - None:  use currently selected armies
    - "custom":  use custom pieces
    - "wall":  use obstacle pieces
  - edit_mode (False):  start in edit mode
  - flip_board (False):  whether to flip the board
  - alter_pieces (0):  when to use alternate sprites for pieces
    - 0:  only in promotion/drop UI
    - 1:  only for white
    - 2:  only for black
    - -1:  always
    - -2:  never
  - alter_swap (False):  whether to swap the side for which alternate sprites are used after a move
  - hide_pieces (0):  piece hiding mode
    - 0:  show all pieces
    - 1:  only show presence (pieceblind mode)
    - 2:  only show type (Penultima mode)
  - hide_moves (None):  whether to hide move markers
    - None:  only if pieces are hidden
    - True/False:  hide/show move markers
  - use_drops (False):  whether piece drops are allowed by default

SEED:  Randomization settings
  - block_ids (None):  army IDs which are prevented from randomly appearing, separated by commas
  - block_ids_chaos (None):  army IDs that are disallowed in chaotic armies, separated by commas
  - chaos_mode (0):  mode for chaotic army generation
    - 0:  only use predefined armies
    - 1:  4 piece types (on matching positions)
    - 2:  7 piece types (on matching positions)
    - 3:  4 piece types (anywhere)
    - 4:  7 piece types (anywhere)
  - chaos_seed (None):  seed for chaotic army generation (if any)
  - set_seed (None):  seed for predefined army selection (if any)
  - roll_seed (None):  seed for probabilistic move types (if any)
  - update_roll_seed (True):  whether to generate new roll seeds when the match is restarted
  - max_seed (2 ** 32 - 1):  maximum seed value to use for randomization

SKIP:  Granular options for automatically making forced moves
  - fast_moves (False):  if there is one legal move, make it
  - fast_chain (True):  same as fast_moves, but only for move chain continuations
  - fast_drops (True):  if only one drop to the selected square is available, make it
  - fast_promotion (True):  if only one promotion of the moving piece is available, make it
  - fast_sequences (True):  same as fast_moves, but only for subsequent moves by the same player
  - fast_turn_pass (True):  if the only legal option is to pass the turn, do so

LOGS:  Logging and status settings
  - log_path ("logs"):  directory to save logs in
  - log_info (True):  whether to log game information on load
  - log_pass (None):  whether to log turn passes
    - None:  only if turn pass was optional
    - True/False:  always/never log
  - log_prefix (1):  type of prefix to use for logging
    - -1:  no prefix
    - 0:  ply count
    - 1:  turn count and player
    - 2:  turn count, player, and move number (per player per turn)
    - 3:  ply count, turn count, player, and move number
  - status_prefix (1):  type of prefix to use for the status string (same as above)
  - status_string (True):  what is shown in the status string
    - True:  show prefix, current game state, and selected move/piece
    - False:  only show prefix, or variant name if no prefix is used
    - None:  only show variant name
  - timestamp (False):  whether to include timestamps in logs
    - True:  always include
    - False:  never include
    - None:  only include if the log is verbose
  - timestamp_format ("%Y-%m-%d %H:%M:%S"):  format for timestamps
  - verbose (True):  whether to use verbose console output by default
    - True:  show all messages
    - False:  only show important messages
    - None:  suppress all messages

SAVE:  Saving and loading settings
  - save_path ("save"):  default directory to store save files in
  - load_path ("save"):  default directory to load save files from
  - load_save (None):  save file to preload (unless defined otherwise by command line)
  - indent (None):  number of spaces to use for indentation in save files
  - compression (0):  level of whitespace compression in save files
    - 0:  no compression (standard JSON output)
    - > 0:  compress elements with depth less than or equal to this value
  - update_mode (0):  how to handle loading a save file
    - 0:  load the save file as is
    - 1:  load the save file and immediately save it again
    - 2:  load the save file, reload the move history, and save it again
    - -1, -2:  same as above, but if the save file was preloaded, exit after saving
  - size_limit (0):  maximum save file size to load (supports suffixes, 0 for no limit)
  - trim_save (False):  whether to remove fields absent from the original save file
  - recursive_aliases (True):  resolution method for aliases in save files
    - True:  resolve aliases recursively
    - False:  resolve aliases once
    - None:  do not resolve aliases

AUTO:  Auto-save settings
  - autosave_path ("auto"):  directory to store auto-saves in
  - autosave_act (0):  how many actions should pass between auto-saves
  - autosave_ply (0):  how many plies should pass between auto-saves
  - autosave_time (0):  how many seconds should pass between auto-saves
  - trim_autosave (False):  whether to remove fields absent from the original save file in auto-saves

SYNC: Server synchronization settings
  - sync_data (False):  whether to synchronize game data with the server
  - sync_host (localhost):  synchronization server address
  - sync_port (58084):  synchronization server port
  - sync_time (0):  how many seconds should pass between idle sync attempts


chess/config.py
from __future__ import annotations

from configparser import ConfigParser
from os.path import isfile
from typing import Any

DEFAULT_CONFIG = {
    "INIT": {
        'asset_path': 'assets',
        'color_id': 0,
        'white_id': 0,
        'black_id': 0,
        'edit_id': '',
        'edit_mode': False,
        'flip_board': False,
        'alter_pieces': 0,
        'alter_swap': False,
        'hide_pieces': 0,
        'hide_moves': '',
        'use_drops': False,
    },
    "SEED": {
        'block_ids': '',
        'block_ids_chaos': '',
        'chaos_mode': 0,
        'chaos_seed': '',
        'set_seed': '',
        'roll_seed': '',
        'update_roll_seed': True,
        'max_seed': 2 ** 32 - 1,
    },
    "SKIP": {
        'fast_moves': False,
        'fast_chain': True,
        'fast_drops': True,
        'fast_promotion': True,
        'fast_sequences': True,
        'fast_turn_pass': True,
    },
    "LOGS": {
        'log_path': 'logs',
        'log_pass': '',
        'log_info': True,
        'log_prefix': 1,
        'status_prefix': 1,
        'status_string': True,
        'timestamp': '',
        'timestamp_format': '%Y-%m-%d %H:%M:%S',
        'verbose': True,
    },
    "SAVE": {
        'save_path': 'save',
        'load_path': 'save',
        'load_save': '',
        'indent': '',
        'compression': 0,
        'update_mode': 0,
        'size_limit': '1M',
        'trim_save': False,
        'recursive_aliases': True,
    },
    "AUTO": {
        'autosave_path': 'auto',
        'autosave_act': 0,
        'autosave_ply': 0,
        'autosave_time': 0,
        'trim_autosave': False,
    },
    "SYNC": {
        'sync_data': False,
        'sync_host': 'localhost',
        'sync_port': 58084,
        'sync_time': 0,
    },
}


class Config(dict):
    def __init__(self, path: str = ''):
        super().__init__()
        self.base_config = None
        self.load(path)

    def load(self, path: str) -> None:
        self.clear()
        self.base_config = ConfigParser(interpolation=None)
        self.base_config.read_dict(DEFAULT_CONFIG)
        if isfile(path):
            self.base_config.read(path)
        for section in self.base_config:
            for item in self.base_config[section]:
                self[item] = self.base_config[section][item]
                try:
                    if type(DEFAULT_CONFIG.get(section, {}).get(item, {})) is bool:
                        self[item] = self.base_config.getboolean(section, item)
                    if type(DEFAULT_CONFIG.get(section, {}).get(item, {})) is int:
                        self[item] = self.base_config.getint(section, item)
                except ValueError:
                    self[item] = DEFAULT_CONFIG[section][item]
                if item in {'hide_moves', 'log_pass', 'status_string', 'timestamp', 'verbose'}:
                    try:
                        self[item] = self.base_config.getboolean(section, item)
                    except ValueError:
                        self[item] = None
                if item.startswith('block_'):
                    self[item] = [
                        int(s) for i in self.base_config[section][item].split(',') if (s := i.strip()).isdigit()
                    ]
                if item.startswith('size_'):
                    self[item] = 0
                    data = self.base_config[section][item].strip().upper()
                    if data[-1:] == 'B':
                        data = data[:-1]
                    exponent = {'K': 10, 'M': 20, 'G': 30, 'T': 40}.get(data[-1:], 0)
                    data = data[:-1] if exponent else data
                    if data.isdigit():
                        self[item] = int(data) << exponent
                if (
                    item == 'indent' or item.endswith('_id')
                    or (item.endswith('_seed') and not item.startswith('update_'))
                ):
                    if self.base_config[section][item].strip() == '':
                        self[item] = None
                    elif self.base_config[section][item].isdigit():
                        self[item] = self.base_config.getint(section, item)
                    else:
                        self[item] = self.base_config[section][item]

    def save(self, path: str) -> None:
        for section in self.base_config:
            for item in self.base_config[section]:
                if self[item] is None:
                    self.base_config[section][item] = ''
                elif item.startswith('block_'):
                    self.base_config[section][item] = ', '.join(str(s) for s in self[item])
                elif item.startswith('size_'):
                    value = self[item]
                    for exponent in ['', 'K', 'M', 'G', 'T']:
                        if value & 1023 == 0:
                            value >>= 10
                        else:
                            break
                    self.base_config[section][item] = f'{value}{exponent}'
                else:
                    self.base_config[section][item] = str(self[item])
        with open(path, mode='w', encoding='utf-8') as file:
            self.base_config.write(file)

    def __copy__(self) -> Config:
        copy = Config()
        for section in self.base_config:
            for item in self.base_config[section]:
                copy.base_config[section][item] = self.base_config[section][item]
        for item in self:
            copy[item] = self[item]
        return copy

    def __deepcopy__(self, memo) -> Config:
        copy = Config()
        for section in self.base_config:
            for item in self.base_config[section]:
                copy.base_config[section][item] = self.base_config[section][item].__deepcopy__(memo)
        for item in self:
            copy[item] = self[item].__deepcopy__(memo)
        return copy

    def __getitem__(self, key: str) -> Any:
        return super().get(key.lower())

    def __setitem__(self, key: str, value: Any) -> None:
        super().__setitem__(key.lower(), value)
